### DISCUSSION TITLE
Some way to force exact path matching

### DISCUSSION THREAD

**Discussion by la55u** [mswjs/msw#1426] [General]
### Scope

Improves an existing behavior

### Compatibility

- [ ] This is a breaking change

### Feature description

I love MSW but recently stumbled upon a behavior that I believed was a bug and spent half a day tracing it down.

As demonstrated here https://github.com/mswjs/msw/issues/814 and here https://github.com/mswjs/msw/issues/742 the order of the handlers have a significant role since the first one that matches will be used. I believe this is not very intuitive as it may cause a lot of headache if you have to sort hundreds of handlers for a large codebase manually.

For example I have two handlers with these path patterns in the following order.

- `/institutes/:id`
- `/institutes/changes`

When I try to make a request to `/institutes/changes` it will give me the response defined in the handler for `/institutes/:id`. Technically it's working because the path matches the pattern but if I want the _correct_ path to match I have to reverse the order which gets very tedious to maintain very quickly especially if the mock handlers are auto generated with another tool. Tools like [orval](https://github.com/anymaniax/orval/issues/606) — and people as seen above  — assume that MSW will just pick up the correct handler without worrying about their order.

So my question is, is it possible currently to force MSW to pick up exact path matches first regardless of the handler's order and then fall back to paths with parameters? If not would it be possible to add a flag to do so?

Thank you

**Comment by kettanaito** [comment]
Hey, @la55u. Thanks for raising this. 

In terms of path matching, MSW is no different from ExpressJS. When there are ambiguous paths like `/movie/latest` and `/movie/:id` the order does become important as there's is virtually no other way a router can know what you mean. It's the same when writing backend route handlers for these two endpoints—the order will matter. Both paths would match `/movie/latest`, so unless there's an additional matching logic (like what type `:id:` should be), the first match will be used. 

You can account for this by structuring the handlers so that more specific handlers come first, more general handlers come last:

- `/movie/latest`
- `/movie/:id`

This is also how any tooling that generates MSW handlers should structure them. 

> So my question is, is it possible currently to force MSW to pick up exact path matches first regardless of the handler's order and then fall back to paths with parameters?

No. Whenever a request happens, MSW goes through each request handler in sequence and uses the first handler that matches the request to produce a mocked response. In this regard, the library is extremely predictable and I wouldn't change that by introducing configurations or implicit behaviors. 

> If not would it be possible to add a flag to do so?

I wouldn't recommend doing this. Right now, the only thing that controls path matching is the `RequestHandler.predicate()` method. I would like for it to stay that way so the logic is predictable and, as the result, extensible. 

---

How would you handle this in Express/other JavaScript backend framework? Perhaps we can get some inspiration from that.

**Reply by la55u** [reply]
@kettanaito 
Sorry I'm not skilled enough to think of a better way to solve this issue. 
But since my question I found [another tool](https://github.com/zoubingwu/msw-auto-mock/issues/14) that doesn't work because of this. :/
